RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1033 @@
+
+```
+List(System.Collections.Generics.IEnumerable<'T>)
+
+    where 'T is (string | null)
+
+Initializes an instance of a [...]
+```
+
+#### Signatures in F# Interactive
+
+To have parity with F# signatures (note: not tooltips), signatures printed in F# interactive will have nullability expression as if it were in code:
+
+```
+> let s: string | null = "hello";;
+
+    val s : string | null = "hello"
+```
+
+#### F# scripting
+
+F# scripts will now also give warnings when using the compiler that implements this feature. This also necessitates a new feature for F# scripting that allows you to opt-out of any nullability warnings. Example:
+
+```fsharp
+#nowarn "nullness"
+
+let s: string = null // OK, since we don't track nullability warnings
+```
+
+By default, F# scripts that use a newer compiler via F# interactive will understand nullability.
+
+#### Project configurability
+
+There are a few concerns here described further in the [Interaction Model](FS-1060-nullable-reference-types.md#interaction-model). When working in an IDE such as visual studio, I may wish to adopt this feature with any of the following configuration toggles:
+
+1. Turn off nullability globally
+2. Turn off nullability on a per-project basis
+3. Turn off nullability for a given assembly (e.g., a package reference)
+4. Turn off warnings for checking of nullable references
+5. Turn off warnings for checking of non-nullable references
+6. Make checking of nullable references an error, separately from `warnaserror`
+7. Make checking of non-nullable references an error, separately from `warnaserror`
+8. Make all nullability warnings an error, separately from `warnaserror`
+
+This implies that there are toggles in the tooling to support this.
+
+* The project properties page for F# projects will require toggles for 2, 4, 5, 6, 7, and 8.
+* There will need to be a right-click gesture that you can use to turn off nullability for a referenced assembly.
+* Turning it off globally is not something that we can control, so (1) is TBD until we can arrive at a tooling solution for all languages.
+
+### Interaction model
+
+The following details how an F# compiler with nullable reference types will behave when interacting with different kinds of components.
+
+#### F# 5.0 consuming non-F# assemblies that do not have nullability
+
+All non-F# assemblies built with a compiler that does not understand nullability are treated such that all of their reference types are nullable.
+
+Example: a `.dll` coming from a NuGet package built with C# 7.1 will be interpreted as if all reference types are nullable, including constraints, generics, etc.
+
+This means that warnings will be emitted when `null` is not properly accounted for. Additionally, when older non-F# components are eventually recompiled with C# 8.0, it will likely be an API-breaking change that will force new F# code to go back and do away with any redundant `null` checks.
```

Comment:
> force new F# code to go back and do away with any redundant `null` checks.

In C#, redundant checks are ok. Consider the scenario where you're implementing a public API consumed by older compilers. Although you declare your method as not accepting `null`, you may still receive `null`.

Reply:
Thanks, updating this to state that the redundancy is fine, but we may wish to consider emitting something letting the user know that it's redundant.